scripts/irods/paths.py

```diff
@@ -1,139 +1,458 @@
+        _scripts_directory_cache = pathlib.Path(os.path.abspath(__file__)).parent.parent
+    return _scripts_directory_cache
+
 def scripts_directory():
-    return os.path.dirname(os.path.dirname(os.path.abspath(
-        inspect.stack()[0][1])))
+    return str(_scripts_directory())
+
+_test_directory_cache = None
+def _test_directory():
+    global _test_directory_cache
+    if _test_directory_cache is None:
+        _test_directory_cache = _scripts_directory() / 'irods' / 'test'
+    return _test_directory_cache
 
 def test_directory():
-    return os.path.join(scripts_directory(), 'irods', 'test')
+    return str(_test_directory())
+
+_server_config_path_cache = None
+def _server_config_path():
+    global _server_config_path_cache
+    if _server_config_path_cache is None:
+        _server_config_path_cache = _config_directory() / 'server_config.json'
+    return _server_config_path_cache
 
 def server_config_path():
-    return os.path.join(
-        config_directory(),
-        'server_config.json')
+    return str(_server_config_path())
+
+_database_config_path_cache = None
+def _database_config_path():
+    global _database_config_path_cache
+    if _database_config_path_cache is None:
+        _database_config_path_cache = _config_directory() / 'database_config.json'
+    return _database_config_path_cache
 
 def database_config_path():
-    return os.path.join(
-        config_directory(),
-        'database_config.json')
+    return str(_database_config_path())
+
+_version_path_cache = None
+def _version_path():
+    global _version_path_cache
+    if _version_path_cache is None:
+        _version_path_cache = _irods_directory() / 'version.json'
+    return _version_path_cache
 
 def version_path():
-    return os.path.join(
-        irods_directory(),
-        'version.json')
+    str(_version_path())
+
+_hosts_config_path_cache = None
+def _hosts_config_path():
+    global _hosts_config_path_cache
+    if _hosts_config_path_cache is None:
+        _hosts_config_path_cache = _config_directory() / 'hosts_config.json'
+    return _hosts_config_path_cache
 
 def hosts_config_path():
-    return os.path.join(
-        config_directory(),
-        'hosts_config.json')
+    return str(_hosts_config_path())
+
+_host_access_control_config_path_cache = None
+def _host_access_control_config_path():
+    global _host_access_control_config_path_cache
+    if _host_access_control_config_path_cache is None:
+        _host_access_control_config_path_cache = _config_directory() / 'host_access_control_config.json'
+    return _host_access_control_config_path_cache
 
 def host_access_control_config_path():
-    return os.path.join(
-        config_directory(),
-        'host_access_control_config.json')
+    return str(_host_access_control_config_path())
 
+# TODO: consider renaming to convey that this is specifically the service account's userconf directory
+# TODO: cache this if we start caching _home_directory()
+def _userconf_directory():
```

Comment:
Same as `_localstate_*`

Reply:
This one is going to change once we start properly following the XDG specifications.
Right now it's just `~/.irods`, but application-specific dotfolders violate the spec, so this will eventually *evaluate to* `~/.config/irods` under normal circumstances, though setting a few environment variables can change the result.

Reply:
Relevant: #5219